File: C/mng.c
#include "mng.h"

#include <stdlib.h>
#include <string.h>

static const uint8_t mng_signature[8] = {0x8a, 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
static const uint8_t png_signature[8] = {0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};

static uint32_t read_u32(const uint8_t *data) {
    return ((uint32_t)data[0] << 24) | ((uint32_t)data[1] << 16) | ((uint32_t)data[2] << 8) | data[3];
}

static int append_bytes(uint8_t **buffer, size_t *length, size_t *capacity, const uint8_t *data, size_t count) {
    if (count > SIZE_MAX - *length) return 0;
    size_t required = *length + count;
    if (required > *capacity) {
        size_t next = *capacity ? *capacity : 64;
        while (next < required) {
            if (next > SIZE_MAX / 2) {
                next = required;
                break;
            }
            next *= 2;
        }
        uint8_t *resized = (uint8_t *)realloc(*buffer, next);
        if (!resized) return 0;
        *buffer = resized;
        *capacity = next;
    }
    memcpy(*buffer + *length, data, count);
    *length = required;
    return 1;
}

static int parse_frame_delay(const uint8_t *data, size_t length, uint32_t *delay) {
    if (length == 0) return 0;
    size_t offset = 1;
    while (offset < length && data[offset] != 0) offset++;
    if (offset == length) return 0;
    offset++;
    if (offset >= length) return 0;
    int has_delay = data[offset++] != 0;
    if (length - offset < 3) return 0;
    offset += 3;
    if (!has_delay || length - offset < 4) return 0;
    *delay = read_u32(data + offset);
    return 1;
}

static int append_frame(mng_animation *animation, uint8_t *png, size_t png_size, uint32_t delay) {
    if (animation->frame_count == SIZE_MAX / sizeof(*animation->frames)) return 0;
    size_t next_count = animation->frame_count + 1;
    mng_frame *frames = (mng_frame *)realloc(animation->frames, next_count * sizeof(*frames));
    if (!frames) return 0;
    animation->frames = frames;
    animation->frames[animation->frame_count].png = png;
    animation->frames[animation->frame_count].png_size = png_size;
    animation->frames[animation->frame_count].delay_milliseconds = delay;
    animation->frame_count = next_count;
    return 1;
}

int mng_is_mng(const uint8_t *data, size_t length) {
    return data != NULL && length >= sizeof(mng_signature) && memcmp(data, mng_signature, sizeof(mng_signature)) == 0;
}

void mng_animation_free(mng_animation *animation) {
    if (!animation) return;
    for (size_t index = 0; index < animation->frame_count; index++) free(animation->frames[index].png);
    free(animation->frames);
    memset(animation, 0, sizeof(*animation));
}

int mng_parse(const uint8_t *data, size_t length, mng_animation *animation) {
    if (!animation || !mng_is_mng(data, length)) return 0;
    memset(animation, 0, sizeof(*animation));
    animation->ticks_per_second = 100;
    uint8_t *frame = NULL;
    size_t frame_size = 0;
    size_t frame_capacity = 0;
    int in_png = 0;
    int has_pending_delay = 0;
    uint32_t pending_delay = 0;
    size_t offset = sizeof(mng_signature);
    while (offset < length) {
        if (length - offset < 12) goto fail;
        uint32_t data_length = read_u32(data + offset);
        size_t chunk_length = (size_t)data_length + 12;
        if (chunk_length < 12 || chunk_length > length - offset) goto fail;
        const uint8_t *chunk_data = data + offset + 8;
        if (memcmp(data + offset + 4, "MHDR", 4) == 0) {
            if (data_length < 28) goto fail;
            animation->width = read_u32(chunk_data);
            animation->height = read_u32(chunk_data + 4);
            uint32_t ticks = read_u32(chunk_data + 8);
            if (ticks > 0) animation->ticks_per_second = ticks;
            animation->nominal_layer_count = read_u32(chunk_data + 12);
            animation->nominal_frame_count = read_u32(chunk_data + 16);
            animation->nominal_play_time = read_u32(chunk_data + 20);
            animation->simplicity_profile = read_u32(chunk_data + 24);
        } else if (memcmp(data + offset + 4, "FRAM", 4) == 0) {
            has_pending_delay = parse_frame_delay(chunk_data, data_length, &pending_delay);
        }
        if (!in_png && memcmp(data + offset + 4, "IHDR", 4) == 0) {
            if (!append_bytes(&frame, &frame_size, &frame_capacity, png_signature, sizeof(png_signature))) goto fail;
            in_png = 1;
        }
        if (in_png) {
            if (!append_bytes(&frame, &frame_size, &frame_capacity, data + offset, chunk_length)) goto fail;
            if (memcmp(data + offset + 4, "IEND", 4) == 0) {
                uint32_t delay = 100;
                if (has_pending_delay) {
                    uint64_t milliseconds = ((uint64_t)pending_delay * 1000 + animation->ticks_per_second / 2) / animation->ticks_per_second;
                    delay = milliseconds > UINT32_MAX ? UINT32_MAX : (uint32_t)milliseconds;
                }
                if (!append_frame(animation, frame, frame_size, delay)) goto fail;
                frame = NULL;
                frame_size = 0;
                frame_capacity = 0;
                has_pending_delay = 0;
                pending_delay = 0;
                in_png = 0;
            }
        }
        offset += chunk_length;
    }
    if (in_png || animation->frame_count == 0) goto fail;
    free(frame);
    return 1;
fail:
    free(frame);
    mng_animation_free(animation);
    return 0;
}

int mng_first_png_frame(const uint8_t *data, size_t length, uint8_t **png, size_t *png_size) {
    if (!png || !png_size) return 0;
    *png = NULL;
    *png_size = 0;
    mng_animation animation = {0};
    if (!mng_parse(data, length, &animation)) return 0;
    *png = (uint8_t *)malloc(animation.frames[0].png_size);
    if (!*png) {
        mng_animation_free(&animation);
        return 0;
    }
    memcpy(*png, animation.frames[0].png, animation.frames[0].png_size);
    *png_size = animation.frames[0].png_size;
    mng_animation_free(&animation);
    return 1;
}

size_t mng_frame_index_at(const mng_animation *animation, uint64_t elapsed_milliseconds) {
    if (!animation || animation->frame_count == 0) return SIZE_MAX;
    if (animation->frame_count == 1) return 0;
    uint64_t duration = 0;
    for (size_t index = 0; index < animation->frame_count; index++) duration += animation->frames[index].delay_milliseconds;
    if (duration == 0) return 0;
    uint64_t elapsed = elapsed_milliseconds % duration;
    for (size_t index = 0; index < animation->frame_count; index++) {
        if (elapsed < animation->frames[index].delay_milliseconds) return index;
        elapsed -= animation->frames[index].delay_milliseconds;
    }
    return animation->frame_count - 1;
}

File: Zig/mng.zig
const std = @import("std");

const mng_signature = [_]u8{ 0x8a, 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a };
const png_signature = [_]u8{ 0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a };

pub const Error = error{ NotMNG, TruncatedChunk, InvalidChunkLength, InvalidHeader, NoFrames };

pub const Frame = struct {
    png: []u8,
    delay_milliseconds: u32,
};

pub const Animation = struct {
    allocator: std.mem.Allocator,
    width: u32 = 0,
    height: u32 = 0,
    ticks_per_second: u32 = 100,
    nominal_layer_count: u32 = 0,
    nominal_frame_count: u32 = 0,
    nominal_play_time: u32 = 0,
    simplicity_profile: u32 = 0,
    frames: []Frame,

    pub fn deinit(self: *Animation) void {
        for (self.frames) |frame| self.allocator.free(frame.png);
        self.allocator.free(self.frames);
    }

    pub fn frameIndexAt(self: Animation, elapsed_milliseconds: u64) ?usize {
        if (self.frames.len == 0) return null;
        if (self.frames.len == 1) return 0;
        var duration: u64 = 0;
        for (self.frames) |frame| duration += frame.delay_milliseconds;
        if (duration == 0) return 0;
        var elapsed = elapsed_milliseconds % duration;
        for (self.frames, 0..) |frame, index| {
            if (elapsed < frame.delay_milliseconds) return index;
            elapsed -= frame.delay_milliseconds;
        }
        return self.frames.len - 1;
    }
};

fn readU32(data: []const u8, offset: usize) u32 {
    return (@as(u32, data[offset]) << 24) | (@as(u32, data[offset + 1]) << 16) | (@as(u32, data[offset + 2]) << 8) | data[offset + 3];
}

fn parseFrameDelay(data: []const u8) ?u32 {
    if (data.len == 0) return null;
    var offset: usize = 1;
    while (offset < data.len and data[offset] != 0) offset += 1;
    if (offset == data.len) return null;
    offset += 1;
    if (offset >= data.len) return null;
    const has_delay = data[offset] != 0;
    offset += 1;
    if (data.len - offset < 3) return null;
    offset += 3;
    if (!has_delay or data.len - offset < 4) return null;
    return readU32(data, offset);
}

pub fn isMNG(data: []const u8) bool {
    return data.len >= mng_signature.len and std.mem.eql(u8, data[0..mng_signature.len], mng_signature[0..]);
}

pub fn parse(allocator: std.mem.Allocator, data: []const u8) !Animation {
    if (!isMNG(data)) return error.NotMNG;
    var frames = std.ArrayList(Frame).init(allocator);
    errdefer {
        for (frames.items) |frame| allocator.free(frame.png);
        frames.deinit();
    }
    var animation = Animation{ .allocator = allocator, .frames = undefined };
    var frame = std.ArrayList(u8).init(allocator);
    defer frame.deinit();
    var pending_delay: ?u32 = null;
    var in_png = false;
    var offset: usize = mng_signature.len;
    while (offset < data.len) {
        if (data.len - offset < 12) return error.TruncatedChunk;
        const data_length = @as(usize, readU32(data, offset));
        const chunk_length = data_length + 12;
        if (chunk_length < 12 or chunk_length > data.len - offset) return error.InvalidChunkLength;
        const chunk_type = data[offset + 4 .. offset + 8];
        const chunk_data = data[offset + 8 .. offset + 8 + data_length];
        if (std.mem.eql(u8, chunk_type, "MHDR")) {
            if (data_length < 28) return error.InvalidHeader;
            animation.width = readU32(chunk_data, 0);
            animation.height = readU32(chunk_data, 4);
            const ticks = readU32(chunk_data, 8);
            if (ticks > 0) animation.ticks_per_second = ticks;
            animation.nominal_layer_count = readU32(chunk_data, 12);
            animation.nominal_frame_count = readU32(chunk_data, 16);
            animation.nominal_play_time = readU32(chunk_data, 20);
            animation.simplicity_profile = readU32(chunk_data, 24);
        } else if (std.mem.eql(u8, chunk_type, "FRAM")) {
            pending_delay = parseFrameDelay(chunk_data);
        }
        if (!in_png and std.mem.eql(u8, chunk_type, "IHDR")) {
            try frame.appendSlice(png_signature[0..]);
            in_png = true;
        }
        if (in_png) {
            try frame.appendSlice(data[offset .. offset + chunk_length]);
            if (std.mem.eql(u8, chunk_type, "IEND")) {
                const delay = if (pending_delay) |ticks| @as(u32, @intCast((@as(u64, ticks) * 1000 + animation.ticks_per_second / 2) / animation.ticks_per_second)) else 100;
                const png = try frame.toOwnedSlice();
                frames.append(.{ .png = png, .delay_milliseconds = delay }) catch |err| {
                    allocator.free(png);
                    return err;
                };
                pending_delay = null;
                in_png = false;
            }
        }
        offset += chunk_length;
    }
    if (in_png or frames.items.len == 0) return error.NoFrames;
    animation.frames = try frames.toOwnedSlice();
    return animation;
}

pub fn firstPngFrame(allocator: std.mem.Allocator, data: []const u8) ![]u8 {
    var animation = try parse(allocator, data);
    defer animation.deinit();
    const result = try allocator.alloc(u8, animation.frames[0].png.len);
    @memcpy(result, animation.frames[0].png);
    return result;
}
